dt: start exposing dependency ordinals

mbolivar-nordic opened the pull request:
Add the first API functions that directly deal with node dependency
ordinals as determined by edtlib:

- DT_DEP_ORD(node_id): node_id's ordinal
- DT_CHILD_DEP_ORD(node_id, child): a particular child's ordinal
- DT_CHILDREN_DEP_ORDS(node_id): an array initializer for a node's
  children's ordinals
- DT_INST_ equivalents

This is not meant to be an exhaustive set of macros related to
dependency ordinals; rather, it's a starting out point meant to enable
initial struct device dependency tracking work. We'll add more as
needed.

Fixes: #26050 (though I bike-shedded the names a bit; feel free to ask for changes there)

mbolivar-nordic committed:
devicetree: add first round of dependency ordinal info

Add the first API functions that directly deal with node dependency
ordinals as determined by edtlib:

- DT_DEP_ORD(node_id): node_id's ordinal
- DT_REQUIRES_DEP_ORDS(node_id): list of dep ordinals for node_id's
  direct dependencies
- DT_SUPPORTS_DEP_ORDS(node_id): list of dep ordinals for nodes
  depending directly on node_id
- DT_INST_ equivalents

This is not meant to be an exhaustive set of macros related to
dependency ordinals; rather, it's a starting out point meant to enable
initial struct device dependency tracking work. We can add more if
needed.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@pabigot I am guessing that at some point you will at least want a macro that expands to an array of a node's direct dependencies' ordinals, but since you didn't ask for that yet, it's not included here.

pabigot commented on include/devicetree.h:
```diff
@@ -1258,6 +1258,62 @@
+ * @return the dependency ordinals of the node's children as a C array
+ *         initializer
+ */
+#define DT_CHILDREN_DEP_ORDS(node_id) \
```
I was unclear about this in part because I was unsure, but I don't think this should be an initializer list, but rather a list, by which I mean I might want to do something like:

    UTIL_LISTIFY(DT_CHILDREN_DEP_ORDS, GEN_HANDLE_SYMBOL);

However, I may be wrong about that.  Early days....

mbolivar-nordic replied:
OK, I made this a list instead of an array. I left the names unchanged, though. I think "DEP_ORD" is easy to understand as a "dependency ordinal". I think that might be a more practical and readable name since English puts adjectives first.

pabigot commented on doc/guides/dts/macros.bnf:
```diff
@@ -50,6 +50,9 @@ node-macro =/ %s"DT_N" path-id %s"_PARENT"
 node-macro =/ %s"DT_N" path-id %s"_STATUS_" dt-name
+; The node's dependency ordinal. This is a non-negative integer
+; value that is used to represent dependency information.
+node-macro =/ %s"DT_N" path-id %s"_DEP_ORD"
```
My original was just `_ORD` because at this time we have only one ordinal associated with each node, which happens to represent one type of dependency information.

If we're adding `DEP` to identify the ordinal type I would prefer to have it be `_ORD_DEP`, reflecting a path from general to specific.  E.g. if we used the dependency ordinal for device power management, but needed a stronger (total) order for device initialization (e.g. incorporating init level), we'd add `_ORD_INIT`.  This also reduces the likelihood of namespace collision.

Just an outline of where my thoughts are right now, no change requested at this time.

pabigot reviewed (commented):
Thanks for this.  Just comments now, because I need to see how all this actually works in practice.  I'll start working with this PR to validate the API design.

mbolivar-nordic commented:
Now updated. Adding DNM because I've left some test cases stubbed out for now. Demonstration which shows how to get nodes with NVIC as interrupt-parent on an Arm Cortex-M:

```

#define NVIC DT_NODELABEL(nvic)

void main(void)
{
	const int sup_ords[] = {
		DT_SUPPORTS_DEP_ORDS(NVIC)
	};
	const char *sup_labels[] = {
		FOR_EACH(DT_LABEL, (,), DT_SUPPORTS(NVIC))
	};
	const size_t n_supported = ARRAY_SIZE(sup_ords);

	printk("the nvic on this SoC supports:\n\n");
	printk("Ordinal Label\n");
	printk("------- ----------\n");
	for (size_t i = 0; i < n_supported; i++) {
		printk("%7d %10s\n", sup_ords[i], sup_labels[i]);
	}
}
```

On qemu_cortex_m3:

```
the nvic on this SoC supports:

Ordinal Label
------- ----------
      8 ETH
      9 GPIO_A
     10 GPIO_B
     11 GPIO_C
     12 GPIO_D
     13 GPIO_E
     14 GPIO_F
     15 GPIO_G
     18 UART_0
     19 UART_1
     20 UART_2
```

pabigot commented:
I have some preliminary experience with this now, and while it's nice and general it doesn't quite suit my desires, and possibly not even my needs.

`FOR_EACH_NONEMPTY` does make this feasible, though I've added #26219 to propose an alternative API that I believe is more flexible.  Yes, I could store every array separately but that not only chews up more space for pointers, I'm also not clear on how I can inject external dependencies for non-device things like sys_init functions (see below) if I can't have a reliable way to ensure the preceding token sequence is syntactically compatible.

More generally it's not clear to me what value we get from `DT_SUPPORTS(node)` producing a sequence of node identifiers, over which `DT_DEP_ORD` must be mapped to get the list of ordinals.

What operation other than `DT_DEP_ORD` would be usefully applied to a set of dependency nodes?

If we had `DT_FOO_DEP_ORDS(node)` based on a generated `DT_FOO_ORDS_nodeid` that expanded to an initializer list of the ordinals (where `FOO` would be `CHILDREN`, `REQUIRES`, and `SUPPORTS`) including terminal comma if non-empty, then I could generate the array described in #26219 fairly easily:

```
static const device_handle_t ordinals[] = {
	DT_DEP_ORD(NODE),
	DT_DEP_ORD(DT_PARENT(NODE)),
	DT_CHILDREN_DEP_ORDS(NODE) /* may be empty so must include terminal */
	DEVICE_HANDLE_SEP,
	DT_REQUIRES_DEP_ORDS(NODE) /* may be empty so must include terminal */
	DRIVER_DEPENDENCIES, /* driver-specific non-device dependences injected at DEVICE_INIT, magic needed for terminal */
	DEVICE_HANDLE_SEP,
	DT_SUPPORTS_DEP_ORDS(NODE) /* may be empty so must include terminal */
	DEVICE_HANDLE_ENDS,
};
```

A couple caveats:
* At this time I don't know if we need all that information; is there a use case for iterating at runtime over children of a device?  What requirements derive from it?
* I'm also not sure yet exactly how the devicetree ordinal will map to the runtime ordinal, since all nodes are given an ordinal but not all nodes correspond to devices.
* I don't think we should get rid of `DT_SUPPORTS`/`DT_REQUIRES` right now but I'm skeptical we're going to need them.

I'm going to proceed as far as I can go with what's already here to evaluate solutions to the first point above, so there's no immediate need to rework anything.  But perhaps this PR should be draft until we have a better idea what's needed.

mbolivar-nordic commented:
> More generally it's not clear to me what value we get from `DT_SUPPORTS(node)` producing a sequence of node identifiers, over which `DT_DEP_ORD` must be mapped to get the list of ordinals.

In general, the node identifier is the basis for these APIs, so exposing them directly hurts nothing and is generally what devicetree.h is built on top of.

> What operation other than `DT_DEP_ORD` would be usefully applied to a set of dependency nodes?

If I wanted to get the names of some dependency related devices and print them out for debugging, I could do so using a for-each macro with DT_LABEL over DT_SUPPORTS/DT_REQUIRES, for instance, just as done in the above examples (though with some problems on empty lists due to the FOR_EACH behavior we've been discussing elsewhere).

> I'm going to proceed as far as I can go with what's already here to evaluate solutions to the first point above, so there's no immediate need to rework anything. But perhaps this PR should be draft until we have a better idea what's needed.

I was all set to implement #26219 on top of https://github.com/zephyrproject-rtos/zephyr/pull/26134 but now I see you're not sure you'll need it yet, so I will hold off. At least we'll get util.h in the API list out of this so far, which is still a win. I'll stand by until you know what you need on the rest. Thanks.

pabigot commented:
If you want to work #26219 I think it's valuable independent of this specific case, and would help get further if you don't want to generate the list of ordinals in the tooling.  There's still some magic required to get the injected non-device dependencies (`DRIVER_DEPENDENCIES` in the example above) to work when they're empty vs non-empty, but that may not be infeasible: we just need an `IS_EMPTY()` macro which I think I know how to do.

mbolivar-nordic commented:
> we just need an `IS_EMPTY()` macro which I think I know how to do.

There is one already; does it not meet your needs?

pabigot commented:
> > we just need an `IS_EMPTY()` macro
> There is one already; does it not meet your needs?

IIRC it didn't work as I expected, but I'll try again later.

mbolivar-nordic commented:
@pabigot I went a slightly different way than your proposal. Rationale in https://github.com/zephyrproject-rtos/zephyr/issues/26219#issuecomment-645739619.

mbolivar-nordic commented:
Also rebased on https://github.com/zephyrproject-rtos/zephyr/pull/26134.

pabigot commented:
> @pabigot I went a slightly different way than your proposal. Rationale in [#26219 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/26219#issuecomment-645739619).

OK.  I responded in #26219 because I saw it there first.  Let's continue to use that issue for that specific change as it can easily be decoupled from dependency-related API.

mbolivar-nordic committed:
doc: document devicetree.h dependency ordinal APIs

Hide these away in the reference documentation for now. It seems
rather unlikely that they'll be used by anything other than internal
infrastructure to get started with.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
After offline discussion, we've removed the dependency on sys/util.h changes for now.

mbolivar-nordic commented:
Rebased to get on top of an unrelated fix for CI purposes.

pabigot commented:
Do you still have `FOR_EACH_NONEMPTY()`?  Or is that in a separate PR somewhere?

I still need that or something like it to handle the dependencies that do not come from devicetree, and are ATT likely to be trailing arguments to the `DEVICE_INIT()` macro, since https://github.com/zephyrproject-rtos/zephyr/pull/26293/files#r443712251 says `IS_EMPTY()` isn't going to work for me.

pabigot commented:
I've confirmed that `P99_IS_EMPTY()` solves the problem I'm having.  I'm not having luck with resurrecting `FOR_EACH_NONEMPTY()` and making it work with a (possibly empty) list of external dependencies.

I think external dependencies will have to be handled through some other means if P99 isn't going to be adoptable in this cycle.

mbolivar-nordic commented:
> Do you still have `FOR_EACH_NONEMPTY()`? Or is that in a separate PR somewhere?

Here's a patch you should be able to `git am`: https://gist.github.com/mbolivar-nordic/32acb6f3939227aca6b862117b6f59e9

mbolivar-nordic commented:
Rebased and pulled out just the pieces @pabigot ended up using in #26616. Ready for review.

mbolivar-nordic commented:
Rebased and pulled out just the pieces @pabigot ended up using in #26616. Ready for review.

mbolivar-nordic commented:
Dev-review 8 October 2020: no feedback 

@galak ping -- this has two approvals, would you be able to take a look?

galak reviewed (commented):
@mbolivar-nordic do we have any other cases of macros in devicetree.h that aren't related directly to properties?  I ask only in that I'm wondering if the namespace of DT_*ORD* macros should be slightly different.

mbolivar-nordic commented:
> @mbolivar-nordic do we have any other cases of macros in devicetree.h that aren't related directly to properties?

Yep, there are a few actually:

- `DT_N_<node_id>_EXISTS`: when a node exists
- `DT_N_<node_id>_BUS_<bus_type>`: node identifier for the bus node of a node
- `DT_N_<node_id_for_a_fixed_partition>_PARTITION_ID`: generated partition identifier number for a subnode of a `fixed-partitions` node
- `DT_N_<node_id>_COMPAT_MATCHES_<some_compat>`: generated for each compatible
- (various other macros related to nodes, not properties, skipped here; see macros.bnf)
- `DT_FOREACH_OKAY_INST_<compat>`: helper macro that our driver foreach macros use
- `DT_COMPAT_<compat>_BUS_<bus-name>`: compatible `compat` has at least one node on bus `bus-name`
- `DT_COMPAT_HAS_OKAY_<compatible>`: compatible `compat` has at least one status "okay" node
- and a couple of others

> related wondering about splitting out the ORD macros into their own header.

Not sure how it's related, but sure, I will do that.

mbolivar-nordic commented:
Ah, I think I see what you're saying. So taking `DT_N_<node_id_for_a_fixed_partition>_PARTITION_ID` for example, that's not directly used in devicetree.h, but is pulled in from `devicetree/fixed-partitions.h`.

The point stands though that several in my list are used directly in devicetree.h, such as `DT_N_<node_id>_EXISTS`, which is used by `DT_NODE_EXISTS()`.

mbolivar-nordic commented:
First force push is just a rebase. Second one moves the new APIs to a new header.

galak reviewed (commented):
did you leave the DT_INST variants in devicetree.h for a reason?

mbolivar-nordic commented:
> did you leave the DT_INST variants in devicetree.h for a reason?

No, that's a mistake, thanks
